### PR TITLE
WIP Nil checking: implementing https://github.com/nim-lang/RFCs/pull/169 [ci-skip]

### PR DESCRIPTION
**Pull request by alehander92**
A rebase with some changes of https://github.com/nim-lang/Nim/pull/10127

### CODE REVIEW

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -1557,6 +1559,7 @@ proc semTypeof2(c: PContext; n: PNode; prev: PType): PType =
   result = t.typ
 
 proc semTypeNode(c: PContext, n: PNode, prev: PType): PType =
+  echo "semTypeNode ", n
```
Comment (Araq) [inline compiler/semtypes.nim]: debug leftover

File: compiler/treetab.nim [hunk 2]
```diff
@@ -12,16 +12,18 @@
...
-  if n == nil: return
+proc hashTree*(n: PNode): Hash =
+  if n.isNil:
+    return
   result = ord(n.kind)
   case n.kind
   of nkEmpty, nkNilLit, nkType:
     discard
   of nkIdent:
+    echo "what", "ident"
```
Comment (Araq) [inline compiler/treetab.nim:23]: debug leftover

File: compiler/nilcheck.nim [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+    history*:  Table[Symbol, seq[History]]
+    previous*: NilMap
+    base*:     NilMap
+    top*:      NilMap
+
+  Nilability* = enum Safe, MaybeNil, Nil
+
+  Check = tuple[nilability: Nilability, map: NilMap]
+
+let resultId = -1
```
Comment (Araq) [inline compiler/nilcheck.nim]: ``const``  instead of ``let`` here.

File: compiler/ast.nim [hunk 4]
```diff
@@ -1331,6 +1331,11 @@ proc newType*(kind: TTypeKind, owner: PSym): PType =
       echo "KNID ", kind
       writeStackTrace()
 
+proc newRefType*(owner: PSym, nilable: bool = false): PType =
```
Comment (juancarlospaco) [inline compiler/ast.nim]:
```suggestion
proc newRefType*(owner: PSym, nilable = false): PType =
```

File: compiler/nilcheck.nim [hunk 5]
```diff
@@ -0,0 +1,622 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2017 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+import ast, astalgo, renderer, ropes, types, intsets, tables, msgs, options, lineinfos, strutils, sequtils, strformat, idents, treetab, hashes
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim:10]:
```suggestion
import 
  ast, astalgo, hashes, idents, intsets, lineinfos, msgs, options, 
  renderer, ropes, sequtils, strformat, strutils, tables, treetab, types
```
Alpha-sorted + line len.

File: compiler/ast.nim (hunk 4 above)
Comment (alehander92) [reply compiler/ast.nim]: huh i am not sure i knew it

Comment (alehander92) [reply compiler/ast.nim]: thanks

File: compiler/nilcheck.nim [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+# call(var arg) # zahary comment
+#   if arg is ref, assume it's MaybeNil after call
+# loop
+#   union of env for 0, 1, 2 iterations as Herb Sutter's paper
+# return
+#   if something: stop (break return etc)
+#   is equivalent to if something: .. else: remain
+# new(ref)
+#   ref becomes Safe
+# each check returns its nilability and map
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim:51]: Why so much commented stuff :grey_question: Kinda bloats the code and is not really doing anything.
Maybe make it into Documentation comments with a `**Note:** Internal compiler documentation.`, or remove it.
:thinking:

Comment (alehander92) [reply compiler/nilcheck.nim:51]: yes, this is just the old code, now we have a spec, i should remove it indeed!

File: compiler/nilcheck.nim [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+    # let b = $n[1]
+    let b = symbol(n[1])
+    # result.map[b] = Safe
+    result.map.store(b, Safe, TAssign, n[1].info)
+    result.nilability = Safe
+  else:
+    result.nilability = typeNilability(n.typ)
+  # echo result.map
+
+proc event(b: History): string =
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim]:
```suggestion
template event(b: History): string =
```
Its just a `case`.

File: compiler/nilcheck.nim [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+  of TType: "it has ref type"
+
+proc derefWarning(n, conf, map; maybe: bool) =
+  var a = history(map, symbol(n))
+  var res = ""
+  res.add("can't deref " & $n & " it " & (if maybe: "might be" else: "is") & " nil")
+  res.add("\n")
+  for b in a:
+    res.add("  " & event(b) & " on line " & $b.info.line & ":" & $b.info.col)
+  message(conf, n.info, warnNilCheck, res)
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim:224]:
```suggestion
  var res = @["can't deref " & $n & " it " & (if maybe: "might be" else: "is") & " nil"]
  for b in a:
    res.add "  " & event(b) & " on line " & $b.info.line & ":" & $b.info.col
  message(conf, n.info, warnNilCheck, res.join"\n")
```
Basically `.add` to 1 `seq[string]` then `.join"\n"` when returning.

File: compiler/nilcheck.nim [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+  result = check(n[0], conf, map)
+  result = checkRefExpr(n, conf, result)
+
+proc checkBracketExpr(n, conf, map): Check =
+  result = check(n[0], conf, map)
+  result = check(n[1], conf, result.map)
+  result = checkRefExpr(n, conf, result)
+
+
+proc union(l: Nilability, r: Nilability): Nilability =
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim]:
```suggestion
template union(l: Nilability, r: Nilability): Nilability =
```
is just 1 `if`.

File: compiler/nilcheck.nim [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+      result.map = map
+  else:
+    result.map = map
+  result.nilability = Safe
+
+proc checkIsNil(n, conf, map; isElse: bool = false): Check =
+  result.map = newNilMap(map)
+  let value = n[1]
+  let value2 = symbol(value)
+  result.map.store(value2, if not isElse: Nil else: Safe, TArg, n.info)
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim]:
```suggestion
  result.map.store(symbol(n[1]), if not isElse: Nil else: Safe, TArg, n.info)
```

`value` and `value2` is only used once and they are short, dont worth the assign of new variable.

File: tests/nilcheck/ref0.nim [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+      b = nil
+  echo b.a
+
+proc f28(b: seq[A]) =
+  for a in b:
+    echo a.a
+
+e(a)
+
+
```
Comment (juancarlospaco) [inline tests/nilcheck/ref0.nim:248]: At least 1 `func` & `template` :grey_question:

Comment (juancarlospaco) [reply tests/nilcheck/ref0.nim:248]: Does this work on `macro` :grey_question: Kinda like:

```nim
macro micro(argument: B not nil): string = 
  newLit"Does this works?. Pseudocode BTW."
```

File: compiler/semtypes.nim [hunk 12]
```diff
@@ -1557,6 +1563,7 @@ proc semTypeof2(c: PContext; n: PNode; prev: PType): PType =
   result = t.typ
 
 proc semTypeNode(c: PContext, n: PNode, prev: PType): PType =
+  # echo "semTypeNode ", n
```
Comment (juancarlospaco) [inline compiler/semtypes.nim]:
```suggestion
```

File: compiler/nilcheck.nim [hunk 13]
```diff
@@ -0,0 +1,624 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2017 Andreas Rumpf
```
Comment (juancarlospaco) [inline compiler/nilcheck.nim:4]:
```suggestion
#        (c) Copyright 2019 Andreas Rumpf
```

File: compiler/nilcheck.nim (hunk 7 above)
Comment (alehander92) [reply compiler/nilcheck.nim]: makes sesne

File: compiler/nilcheck.nim (hunk 8 above)
Comment (alehander92) [reply compiler/nilcheck.nim:224]: good idea! but maybe allocations ? it is a hotspot

File: compiler/nilcheck.nim (hunk 9 above)
Comment (alehander92) [reply compiler/nilcheck.nim]: good idea!

File: compiler/nilcheck.nim (hunk 10 above)
Comment (alehander92) [reply compiler/nilcheck.nim]: no need for value2

File: tests/nilcheck/ref0.nim (hunk 11 above)
Comment (alehander92) [reply tests/nilcheck/ref0.nim:248]: func would be a good idea

template and macro dont make too much sense imo as the analysis is done later when they are expanded : still one can test that

File: compiler/semtypes.nim (hunk 12 above)
Comment (alehander92) [reply compiler/semtypes.nim]: should be erased

File: compiler/nilcheck.nim (hunk 13 above)
Comment (alehander92) [reply compiler/nilcheck.nim:4]: Hm, maybe , not sure how this is maintained

**Comment by alehander92** [comment]
Ok, converting to draft: contains some WIP changes now to the other default, TODO finish RFC first

**Comment by alehander92** [comment]
new PR for the updated new RFC https://github.com/nim-lang/Nim/pull/15287

**Pull request closed without merging**